pdns/reczones-helpers.cc

```diff
@@ -27,115 +27,129 @@
-  dr.d_content = DNSRecordContent::mastermake(QType::SOA, 1, "localhost. root 1 604800 86400 2419200 604800");
-
-  ad.d_records.insert(dr);
 
-  dr.d_type = QType::NS;
-  dr.d_content = std::make_shared<NSRecordContent>(DNSName("localhost."));
+  SyncRes::AuthDomain ad = makeSOAAndNSNodes(dr, DNSName("localhost."));
 
-  ad.d_records.insert(dr);
+  // Go over the hostname and aliases (parts[1], parts[2], etc...) and add PTR entries for
+  // reverse lookups.
   dr.d_type = QType::PTR;
-
-  if (ipParts.size() == 4) { // otherwise this is a partial zone
-    for (unsigned int n = 1; n < parts.size(); ++n) {
-      dr.d_content = DNSRecordContent::mastermake(QType::PTR, 1, DNSName(parts[n]).toString()); // XXX FIXME DNSNAME PAIN CAN THIS BE RIGHT?
-      ad.d_records.insert(dr);
-    }
+  for (auto name = parts.cbegin() + 1; name != parts.cend(); ++name) {
+    dr.d_content = DNSRecordContent::mastermake(QType::PTR, 1, DNSName(*name).toString());
+    ad.d_records.insert(dr);
   }
 
-  if (newMap->count(dr.d_name) != 0) {
-    SLOG(g_log << Logger::Warning << "Will not overwrite zone '" << dr.d_name << "' already loaded" << endl,
-         log->info(Logr::Warning, "Will not overwrite already loaded zone", "zone", Logging::Loggable(dr.d_name)));
-  }
-  else {
-    if (ipParts.size() == 4) {
-      SLOG(g_log << Logger::Warning << "Inserting reverse zone '" << dr.d_name << "' based on hosts file" << endl,
-           log->info(Logr::Notice, "Inserting reverse zone based on hosts file", "zone", Logging::Loggable(dr.d_name)));
-    }
-    ad.d_name = dr.d_name;
-    (*newMap)[ad.d_name] = ad;
+  addToDomainMap(newMap, ad, dr.d_name, log, false, true);
+}
+
+void makePartialIPZone(SyncRes::domainmap_t& newMap,
+                       std::initializer_list<const char*> labels,
+                       Logr::log_t log)
+{
+  DNSRecord dr;
+  for (auto label = std::rbegin(labels); label != std::rend(labels); ++label) {
+    dr.d_name.appendRawLabel(*label);
   }
+  dr.d_name.appendRawLabel("in-addr");
+  dr.d_name.appendRawLabel("arpa");
+
+  SyncRes::AuthDomain ad = makeSOAAndNSNodes(dr, DNSName("localhost."));
+
+  addToDomainMap(newMap, ad, dr.d_name, log, true, true);
 }
 
-void addForwardAndReverseLookupEntries(const std::shared_ptr<SyncRes::domainmap_t>& newMap,
+void addForwardAndReverseLookupEntries(SyncRes::domainmap_t& newMap,
                                        const std::string& searchSuffix,
                                        const std::vector<std::string>& parts,
                                        Logr::log_t log)
 {
-  for (unsigned int n = 1; n < parts.size(); ++n) {
-    if (searchSuffix.empty() || parts[n].find('.') != string::npos) {
-      makeNameToIPZone(newMap, DNSName(parts[n]), parts[0], log);
+  const ComboAddress address{parts[0]};
+
+  // Go over the hostname and aliases (parts[1], parts[2], etc...) and add entries
+  // for forward lookups.
+  for (auto name = parts.cbegin() + 1; name != parts.cend(); ++name) {
```

Comment:
`const auto& name`

Reply:
I must be missing something here. Doesn't the `const` in your suggestion apply to the iterator object here? How would the iterator advance if it cannot be mutated?

Reply:
oh right it's iterator... my bad.